kratos/processes/find_global_nodal_neighbours_for_entities_process.cpp

```diff
@@ -0,0 +1,304 @@
+    });
+
+    // adding the neighbouring nodes
+    if (!this->mrDataCommunicator.IsDistributed()) {
+        BlockPartition<TContainerType>(this->GetContainer())
+            .for_each([&](typename TContainerType::value_type& rEntity) {
+                auto& r_geometry = rEntity.GetGeometry();
+                for (unsigned int i = 0; i < r_geometry.size(); ++i) {
+                    for (unsigned int j = 0; j < r_geometry.size(); ++j) {
+                        if (j != i) {
+                            auto gp = GlobalPointer<NodeType>(r_geometry(j), 0);
+                            auto& r_node = r_geometry[i];
+                            r_node.SetLock();
+                            AddUniqueGlobalPointer<NodeType>(
+                                r_node.GetValue(mrOutputVariable), gp);
+                            r_node.UnSetLock();
+                        }
+                    }
+                }
+            });
+
+        BlockPartition<NodesContainerType>(r_nodes).for_each([&](NodeType& rNode) {
+            auto& r_neighbours = rNode.GetValue(mrOutputVariable);
+            r_neighbours.shrink_to_fit();
+            std::sort(r_neighbours.ptr_begin(), r_neighbours.ptr_end(),
+                      [](GlobalPointer<NodeType> const& gp1,
+                         GlobalPointer<NodeType> const& gp2) {
+                          return gp1->Id() < gp2->Id();
+                      });
+        });
+
+    } else { // mpi case!
+        const int current_rank = this->mrDataCommunicator.Rank();
+
+        using map_of_sets = std::unordered_map<int, std::vector<int>>;
+        std::unordered_map<int, map_of_sets> neighbours_ids;
+
+        for (auto& r_entity : this->GetContainer()) {
+            const auto& r_geometry = r_entity.GetGeometry();
+            for (unsigned int i = 0; i < r_geometry.size(); ++i) {
+                const int i_owner_rank =
+                    r_geometry[i].FastGetSolutionStepValue(PARTITION_INDEX);
+                auto& container = neighbours_ids[i_owner_rank][r_geometry[i].Id()];
+                for (unsigned int j = 0; j < r_geometry.size(); ++j) {
+                    if (j != i) {
+                        AddUnique(container, r_geometry[j].Id());
+                    }
+                }
+            }
+        }
+
+        // there are some isolated nodes when metis partitioner performs partitioning
+        // specially in the case where TContainerType = ModelPart::ConditionsContainerType
+        // therefore we add those ids to neighbours_ids so that proper communication scheduling
+        // can be computed.
+        BlockPartition<NodesContainerType>(r_nodes).for_each([&](const NodeType& rNode) {
+            const int i_owner_rank = rNode.FastGetSolutionStepValue(PARTITION_INDEX);
+            const int node_id = rNode.Id();
+            if (neighbours_ids[i_owner_rank].find(node_id) ==
+                neighbours_ids[i_owner_rank].end()) {
+                #pragma omp critical
```

Comment:
@RiccardoRossi for my own interest, how are you planning to do this in the future with C++ threads? Maybe by using `std::mutex` and `std::lock_guard`?

Reply:
either a global mutex or a custom global spin lock

[spin_lock](https://en.cppreference.com/w/cpp/atomic/atomic_flag)

Reply:
the [LockObject](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/includes/lock_object.h) would also work for this right?